### PR TITLE
Add UPnP view and new port scanning API

### PR DESCRIPTION
**Pull request by pablomendezroyo**
### UPnP common issues

UPnP is one of the most critical features in DAppNode. We have poor control over routers, using the library miniUPnPC, since every router is different and its behavior may change from one to another.

This situation generates a lot of support and chaos in users:

- UPnP is disabled, and users do not even how to enable it.
- UPnP is disabled and the router does not even have this property. Users should have something in the UI advertising to them which ports are necessary and for what.
- UPnP is enabled but dappnode cannot open all the necessary ports, so users do not even know which ports need to be manually opened and the status of these ports
- UPnP is enabled and users want to know which ports and their status of those ones managed by UPnP.

**If UPnP is enabled, and any port is manually closed it will automatically be re-opened**

### Main changes

- Used an already created call `getPortStatus` which have useful information about UPnP and their ports status
- Added new view at **support > UPnP**,  there are 3 main possibilities
     - DAppNode running in a remote machine. UPnP is not necessary
     - DAppNode running locally, UPnP detected as enabled
     - DAppNode running locally, UPnP detected as disabled

- Added API call to scan TCP ports and show their status 

**Results**

**I could not recreate the situation of having UpNp enabled and some ports closed, since these ports are automatically opened by dappmanager**

**DAppNode running in remote**
image

**UpNp enabled and ports succesfully mapped**
image

**UpNp disabled and ports to be open shown**

image

**Scanning**

image

### CODE REVIEW

File: packages/admin-ui/src/pages/support/components/UPnP.tsx [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import ErrorView from "../../../components/ErrorView";
+import SubTitle from "components/SubTitle";
+import Ok from "components/Ok";
+import Card from "components/Card";
+
+function PortsOpened({ localIp }: { localIp: string }) {
+  const upnpInfo = useApi.getPortsStatus();
+
+  return (
+    <>
```
Comment (dapplion) [inline packages/admin-ui/src/pages/support/components/UPnP.tsx:14]: It's not necessary to have this initial React.Fragment

File: packages/admin-ui/src/pages/support/components/UPnP.tsx [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    </>
+  );
+}
+
+function PortsToOpen() {
+  const upnpInfo = useApi.getPortsStatus();
+  const portsScanInfo = useApi.portsScanGet();
+
+  return (
+    <>
```
Comment (dapplion) [inline packages/admin-ui/src/pages/support/components/UPnP.tsx:70]: It's not necessary to have this initial React.Fragment

File: packages/dappmanager/src/utils/getPortsScan.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+export async function getPortsScan({
+  publicIp,
+  ports
+}: {
+  publicIp: string;
+  ports: string[];
+}): Promise<PortsScanResponse> {
+  try {
+    const response = await fetch(
+      `https://mighty-escarpment-70069.herokuapp.com/${publicIp}/${ports.join(
```
Comment (dapplion) [inline packages/dappmanager/src/utils/getPortsScan.ts:13]: Move the url to params.ts as `PORT_SCANNER_SERVICE_URL`

File: packages/admin-ui/src/pages/support/components/UPnP.tsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      ) : (
+        <Loading steps={["Loading opened ports"]} />
+      )}
+    </>
+  );
+}
+
+function PortsToOpen() {
+  const upnpInfo = useApi.getPortsStatus();
+  const portsScanInfo = useApi.portsScanGet();
```
Comment (dapplion) [inline packages/admin-ui/src/pages/support/components/UPnP.tsx:67]: A port scan should be triggered manually by the user, not automatically

File: packages/admin-ui/src/pages/support/components/UPnP.tsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+                          ) : (
+                            <Ok ok={false} msg={"Closed"} />
+                          )
+                        ) : (
+                          <Ok ok={false} msg={"Unknown"} />
+                        )}
+                      </td>
+                    </tr>
+                  );
+                })}{" "}
```
Comment (dapplion) [inline packages/admin-ui/src/pages/support/components/UPnP.tsx:52]: Unnecessary `" "`

File: packages/admin-ui/src/common/types.ts [hunk 6]
```diff
@@ -441,6 +448,14 @@ export type InstalledPackageData = Pick<
   containers: PackageContainer[];
 };
 
+export interface PortsScanResponse {
+  host: string; //"83.32.61.31"
+  ports: {
+    open: number[]; // [8092,4001]
```
Comment (dapplion) [inline packages/admin-ui/src/common/types.ts:454]: Even if we only scan TCP, this information should be clear here. The array should be of type `{port: number, protocol: "TCP" | "UDP" }`

Comment (dapplion) [reply packages/admin-ui/src/common/types.ts:454]: It should return an array of all the ports requested with the status as a property, instead of sorted in two arrays.
```ts
ports: {
  port: number;
  protocol: "TCP" | "UDP";
  status: "open" | "closed" | "error";
  error?: string;
}[]
```

File: packages/admin-ui/src/pages/support/components/UPnP.tsx [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+      ) : upnpInfo.error ? (
+        <ErrorView error={upnpInfo.error} />
+      ) : (
+        <Loading steps={["Loading opened ports"]} />
+      )}
+    </>
+  );
+}
+
+function PortsToOpen() {
```
Comment (dapplion) [inline packages/admin-ui/src/pages/support/components/UPnP.tsx:65]: I think you could merge `PortsOpened` and `PortsToOpen` into the same component and reduce code duplication

File: packages/dappmanager/src/calls/portsScanGet.ts [hunk 8]
```diff
@@ -1,11 +1,11 @@
...
+import { PortScanResponse } from "../types";
 import { getPortsScan } from "../utils/getPortsScan";
 import * as db from "../db";
 
-export async function portsScanGet(): Promise<PortsScanResponse> {
+export async function portsScanGet(): Promise<PortScanResponse[]> {
   const packagePort = db.portsToOpen.get();
   // udp ports cannot be scanned
-  // as a better approach, if tcp ports were opened udp quite probably were opened as well
+  // as a better approach, if tcp ports were opened with UpNp, udp quite probably were opened as well
```
Comment (dapplion) [inline packages/dappmanager/src/calls/portsScanGet.ts:8]: UPnP capitalization

File: packages/dappmanager/src/utils/getPortsScan.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import fetch from "node-fetch";
+import params from "../params";
+import { PortScanResponse } from "../types";
+
+const apiEndpoint = params.PORT_SCANNER_SERVICE_URL;
+
+export async function getPortsScan({
```
Comment (dapplion) [inline packages/dappmanager/src/utils/getPortsScan.ts:7]: Rename to `performPortsScan`

File: packages/dappmanager/src/calls/portsScanGet.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { PortScanResponse } from "../types";
+import { getPortsScan } from "../utils/getPortsScan";
+import * as db from "../db";
+
+export async function portsScanGet(): Promise<PortScanResponse[]> {
+  const packagePort = db.portsToOpen.get();
+  // udp ports cannot be scanned
+  // as a better approach, if tcp ports were opened with UpNp, udp quite probably were opened as well
+  const tcpPorts = packagePort.filter(port => port.protocol === "TCP");
```
Comment (dapplion) [inline packages/dappmanager/src/calls/portsScanGet.ts:9]: This formating should be done in the utility `getPortsScan` to keep things together. This way you can use a fully typed interface instead of `ports: string[]` which actually means only the TCP port number in string form

File: packages/dappmanager/src/utils/getPortsScan.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  ports
+}: {
+  publicIp: string;
+  ports: string[];
+}): Promise<PortScanResponse[]> {
+  try {
+    const response = await fetch(
+      `${apiEndpoint}/${publicIp}?tcpPorts=${ports.join(",")}`
+    );
+    return await response.json();
```
Comment (dapplion) [inline packages/dappmanager/src/utils/getPortsScan.ts:18]: You should sanitize the response from the API before sending it to the UI, just in case

**Comment by pablomendezroyo** [comment]
## PR update

- Added API response sanitize.
- Added service name to the port table
- Added advanced mode: show API and UPnP port status
- Added logic to determine when to take API or UPnP port status (use to be API
- Added logic to build the necessary data for the ports table in the backend

upnpEnabledOn
upnpDisabledOff
upnpDisabledOn
upnpEnabledOff

File: packages/dappmanager/src/utils/portsTableData.ts [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+  const upnpAvailable = db.upnpAvailable.get();
+
+  // COOKING DATA
+  return portsToOpen.map(port => {
+    return {
+      port: port.portNumber,
+      protocol: port.protocol,
+      upnpStatus: getUpnpStatus({ port, upnpAvailable, upnpPortMappings }),
+      apiStatus: getApiStatus({ port, apiTcpPortsStatus }),
+      service: getServiceName({ port, packages })
```
Comment (dapplion) [inline packages/dappmanager/src/utils/portsTableData.ts:34]: You should use strictly one this "name" entities: `dnpName`, `serviceName`, `containerName`. Otherwise things get confusing. Note that `serviceName` does not uniquely identify a service globally. You should return `dnpName`, `serviceName` from the dappmanager and the UI should print that in a pretty way

**Review by dapplion — CHANGES_REQUESTED** [review]

**Review by dapplion — CHANGES_REQUESTED** [review]

File: packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx [hunk 13]
```diff
@@ -1,13 +1,12 @@
-import { useApi } from "api";
 import Button from "components/Button";
 import ErrorView from "components/ErrorView";
 import Loading from "components/Loading";
 import Ok from "components/Ok";
 import SubTitle from "components/SubTitle";
 import Switch from "components/Switch";
 import React, { useState } from "react";
+import { useApi } from "api";
 import { Table } from "react-bootstrap";
```
Comment (dapplion) [inline packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx:9]: You should never import react-bootstrap components from the root to ensure tree-shaking works well. See https://react-bootstrap.github.io/getting-started/introduction/#importing-components

File: packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx [hunk 14]
```diff
@@ -42,9 +24,10 @@ export function PortsStatusTable() {
       <SubTitle>
         Ports table
         <Button
+          disabled={upnpInfo.isValidating ? true : false}
```
Comment (dapplion) [inline packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx:27]: `isValidating` is already a boolean

**Review by dapplion — CHANGES_REQUESTED** [review]

File: packages/dappmanager/src/modules/portsTable/getApiStatus.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
+import { PackagePort, TcpPortScan, ApiStatus } from "../../common";
```
[the remark this answers is not in the document]
Comment (pablomendezroyo) [reply packages/dappmanager/src/modules/portsTable/getApiStatus.ts:1]: donde, only **PortToOpen** is imported from types in the dappmanager

File: packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+
+  async function apiButtonOnClick() {
+    if (portsToOpen.data)
+      try {
+        setApiReqStatus({ loading: true });
+        const apiPorts = await api.portsApiStatusGet({
+          portsToOpen: portsToOpen.data
+        });
+        setApiScanResult(apiPorts);
+        setApiReqStatus({ result: apiPorts });
```
Comment (dapplion) [inline packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx:42]: The result of the api call is stored twice. Remove useState for `setApiScanResult` and `setUpnpScanResult`

File: packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+          portsToOpen: portsToOpen.data
+        });
+        setUpnpScanResult(upnpPorts);
+        setUpnpReqStatus({ result: upnpPorts });
+      } catch (e) {
+        setUpnpReqStatus({ error: e });
+      }
+  }
+
+  function RenderApiStatus({
```
Comment (dapplion) [inline packages/admin-ui/src/pages/support/components/PortsStatusTable.tsx:62]: You should not declare components inside components unless strictly necessary. Move them to the body of the module

File: packages/dappmanager/src/modules/portsTable/performPortsScan.ts [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+      };
+    });
+  }
+}
+
+/**
+ * Check if the response from the scan ports API
+ * is well formatted
+ */
+function sanitizeApiResponse(responseJson: PortScanResponse): void {
```
[the remark this answers is not in the document]
Comment (dapplion) [reply packages/dappmanager/src/modules/portsTable/performPortsScan.ts:50]: @pablomendezroyo Bump

File: packages/dappmanager/src/daemons/natRenewal/index.ts [hunk 19]
```diff
@@ -25,6 +26,7 @@ async function natRenewal(): Promise<void> {
     // portMappings = [ {protocol: 'UDP', exPort: '500', inPort: '500'} ]
     try {
       const portMappings = await upnpc.list();
+
```
Comment (dapplion) [inline packages/dappmanager/src/daemons/natRenewal/index.ts:29]: Unnecessary diff

**Review by dapplion — APPROVED** [review]

**Review by dapplion — APPROVED** [review]
